saw-core/prelude/Prelude.sawcore

```diff
@@ -793,20 +793,360 @@ maybe : (a b : sort 0) -> b -> (a -> b) -> Maybe a -> b;
+   (\ (y : Pos) (_ : Bool -> Bool) (k : Bool) -> rx y True)
+  )
+  (\ (x : Pos) (rx : Pos -> Bool -> Bool) ->
+   Pos_cases (Bool -> Bool)
+   (\ (k : Bool) -> False)
+   (\ (y : Pos) (_ : Bool -> Bool) (k : Bool) -> rx y False)
+   (\ (y : Pos) (_ : Bool -> Bool) (k : Bool) -> rx y k)
+  );
+
+posLe : Pos -> Pos -> Bool;
+posLe x y = posCmp x y True;
+
+posLt : Pos -> Pos -> Bool;
+posLt x y = posCmp x y False;
+
 --------------------------------------------------------------------------------
 -- Nat
 
 data Nat : sort 0 where {
     Zero : Nat;
-    Succ : Nat -> Nat;
+    NatPos : Pos -> Nat;
   }
 
-Nat__rec : (p : Nat -> sort 1) ->
-           p Zero ->
-           ((n:Nat) -> p n -> p (Succ n)) ->
-           (n:Nat) ->
-           p n;
-Nat__rec p f1 f2 n = Nat#rec1 p f1 f2 n;
+Succ : Nat -> Nat;
+Succ n = NatPos (Nat#rec (\ (_ : Nat) -> Pos) One posInc n);
+
+-- | Addition
+addNat : Nat -> Nat -> Nat;
+addNat =
+  Nat#rec (\ (_ : Nat) -> Nat -> Nat)
+  (\ (y : Nat) -> y)
+  (\ (x : Pos) ->
+   Nat#rec (\ (_ : Nat) -> Nat)
+   (NatPos x)
+   (\ (y : Pos) -> NatPos (posAdd x y))
+  );
+
+-- | Multiplication
+mulNat : Nat -> Nat -> Nat;
+mulNat =
+  Nat#rec (\ (_ : Nat) -> Nat -> Nat)
+  (\ (y : Nat) -> Zero)
+  (\ (x : Pos) ->
+   Nat#rec (\ (_ : Nat) -> Nat)
+   Zero
+   (\ (y : Pos) -> NatPos (posMul x y))
+  );
+
+-- | Natural exponentiation
+expNat : Nat -> Nat -> Nat;
+expNat =
+  Nat#rec (\ (_ : Nat) -> Nat -> Nat)
+  (\ (_ : Nat) -> Zero)
+  (\ (x : Pos) ->
+   Nat#rec (\ (_ : Nat) -> Nat)
+   (NatPos One)
+   (\ (y : Pos) -> NatPos (posExp x y))
+  );
+
+equalNat : Nat -> Nat -> Bool;
+equalNat =
+  Nat#rec (\ (_ : Nat) -> Nat -> Bool)
+  (Nat#rec (\ (_ : Nat) -> Bool) True (\ (y : Pos) -> False))
+  (\ (x : Pos) -> Nat#rec (\ (_ : Nat) -> Bool) False (posEq x));
+
+ltNat : Nat -> Nat -> Bool;
+ltNat =
+  Nat#rec (\ (_ : Nat) -> Nat -> Bool)
+  (Nat#rec (\ (_ : Nat) -> Bool) False (\ (y : Pos) -> True))
+  (\ (x : Pos) -> Nat#rec (\ (_ : Nat) -> Bool) False (posLt x));
+
+leNat : Nat -> Nat -> Bool;
+leNat =
+  Nat#rec (\ (_ : Nat) -> Nat -> Bool)
+  (Nat#rec (\ (_ : Nat) -> Bool) True (\ (y : Pos) -> True))
+  (\ (x : Pos) -> Nat#rec (\ (_ : Nat) -> Bool) False (posLe x));
+
+-- | Minimum
+minNat : Nat -> Nat -> Nat;
+minNat x y = ite Nat (ltNat x y) x y;
+
+-- | Maximum
+maxNat : Nat -> Nat -> Nat;
+maxNat x y = ite Nat (ltNat x y) y x;
+
+-- | Width(n) = 1 + floor(log_2(n))
+widthNat : Nat -> Nat;
+widthNat =
+  Nat#rec (\ (_ : Nat) -> Nat)
+  Zero
+  (Pos#rec (\ (_ : Pos) -> Nat)
+   (Succ Zero)
+   (\ (_ : Pos) -> Succ)
+   (\ (_ : Pos) -> Succ)
+  );
+
+--------------------------------------------------------------------------------
+-- Subtraction on Nat
+
+data Z : sort 0 where {
+    ZZero : Z;
+    ZPos : Pos -> Z;
+    ZNeg : Pos -> Z;
+  }
+
+Z_cases : (a : sort 0) -> a -> (Pos -> a) -> (Pos -> a) -> Z -> a;
+Z_cases a = Z#rec (\ (_ : Z) -> a);
+
+-- double and decrement
+-- BitM One = One
```

Comment:
Dumb question: how does this function implement "double and decrement"? `BitM One = One` suggests that it's neither doubling nor decrementing...

Reply:
It computes the function `\n -> 2*n-1`

Reply:
Oops, I have no idea what I was thinking when I wrote this comment. Sorry for the noise!